README.md
# Get Custom Field Values

A plugin for WordPress that provides widgets, shortcodes, and template tags to easily retrieve and display custom field values for posts or pages.

This plugin can be found in the WordPress Plugin Directory: https://wordpress.org/plugins/get-custom-field-values/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Optional: Add filters for `the_meta` to filter custom field data (see the end of the plugin file for commented out samples you may wish to include). And/or add per-meta filters by hooking `the_meta_$field`
4. Give post(s) a custom field with a value.
5. Optional: Go to the Appearance -> Widgets admin page to create one or more 'Get Custom Field' sidebar widgets for your widget-enabled theme.
6. Optional: Use one of the six template functions provided by this plugin to retrieve the contents of custom fields. You must 'echo' the result if you wish to display the value on your site.


## Additional Documentation

* See [readme.txt](https://github.com/coffee2code/get-custom-field-values/blob/master/readme.txt) for additional usage information.
* See [DEVELOPER-DOCS.md](DEVELOPER-DOCS.md) for developer-related documentation on hooks, template tags, and shortcode.
* See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/get-custom-field-values/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/get-custom-field-values/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/get-custom-field-values/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# ChangeLog

## 4.1 _(2023-10-08)_

### Highlights:

This minor release prevents use of unsafe markup in widgets, improves documentation, improves unit testing, and notes compatibility through WP 6.3+.

### Details:

* Hardening: Restrict markup used in widget to safe HTML. Props Satoo Nakano (Patchstack).
* Change: Ensure `show_metabox()` only potentially returns true for post screens
* Change: Add inline documentions for shortcode-related hooks
* Change: Tweak some formatting in `DEVELOPER-DOCS.md`
* Change: Clarify references to shortcode-related hook in `DEVELOPER-DOCS.md`
* Change: Note compatibility through WP 6.3+
* Change: Update copyright date (2023)
* New: Add `.gitignore` file
* Unit tests:
    * Fix: Allow tests to run against current versions of WordPress
    * New: Add `composer.json` for PHPUnit Polyfill dependency
    * Change: Prevent PHP warnings due to missing core-related generated files
    * Change: In bootstrap, add backcompat for PHPUnit pre-v6.0
    * Fix: Fix tests associated with `show_metabox()` and add two new tests

## 4.0.1 _(2021-11-05)_
* Change: Further restrict usage of shortcode to only those with 'unfiltered_html' capability. Props Erwan.
* New: Add a few more possible TODO items

## 4.0 _(2021-11-04)_

### Highlights:

This recommended release prevents users who can't publish posts from using the shortcode in posts (security hardening), adds some new filters, adds DEVELOPER-DOCS.md, notes compatibility through WP 5.8+, and reorganizes and improves unit tests.

### Details:

* Change: Prevent users who can't publish posts from using the shortcode in posts. Props Francesco Carlucci.
    * Hardens security to prevent potentail information disclosure or XSS by authors with limited privileges
    * New: Add shortcode class function `can_author_use_shortcodes()`
    * New: Add filter `'get_custom_field_values/can_author_use_shortcodes'`
    * Change: Prevent shortcodes created by users who cannot publish posts from being evaulated
    * Change: Prevent display of shortcode builder metabox to users who cannot publish posts
* New: Add filter `'get_custom_field_values/show_metabox'` to customize if shortcode builder metabox is shown
* New: Add DEVELOPER-DOCS.md and move template tag and shortcode documentation into it
* Change: Note compatibility through WP 5.8+
* Change: Update copyright date (2021)
* Change: Tweak installation instruction
* Unit tests:
    * Change: Split shortcode-related tests out into their own file
    * Change: Split widget-related tests out into their own file
    * New: Add helper functions to facilitate creating users
        * New: Add `create_user()` for creating a user and optionally making them the current user
        * New: Add `unset_current_user()` for unsetting the current user
        * New: Add `tearDown()` to ensure current user gets unset after each test
    * New: Add unit tests for `show_metabox()`
    * Change: Reduce likelihood of particular randomization tests from failing due to reasonable possibility of subsequent randomization choosing the same item
    * Change: Add optional arg `$make_global` (defaulted to false) to `create_post_with_meta()` to facilitate making the created post global
    * Change: Restructure unit test file structure
        * Change: Move `phpunit/bin/` to `tests/bin/`
        * Change: Move `phpunit/bootstrap.php` into `tests/phpunit/`
        * Change: Move tests from `phpunit/tests/` to `tests/phpunit/tests/`
        * Change: In bootstrap, store path to plugin file constant so its value can be used within that file and in test file
        * Change: In bootstrap, check for test installation in more places and exit with error message if not found
        * Change: Remove 'test-' prefix from unit test files
* New: Add a few more possible TODO items

## 3.9.4 _(2020-09-11)_
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.5+

## 3.9.3 _(2020-05-23)_
* New: Add TODO.md and move existing TODO list from top of main plugin file into it
* Change: Update shortcode builder widget to 007:
    * New: Store object instantiated during `register()`
    * Change: Cast return value of `c2c_get_custom_field_values_post_types` filter as an array
    * Change: Sanitize strings used in markup attributes (hardening)
    * Change: Add a missing textdomain for string translation
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests, and removed commented-out code
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Fix typo in FAQ

## 3.9.2 _(2019-12-08)_
* Change: Note compatibility through WP 5.3+
* Change: Unit test: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Use full URL for readme.txt link to full changelog
* Change: Add link to plugin's page in Plugin Directory to README.md
* Change: Update copyright date (2020)

## 3.9.1 _(2019-03-15)_
* Change: Update shortcode builder widget to 006:
    * Fix: Check that `is_block_editor()` exists before using it, preventing an error when attempting to edit pages in WP earlier than 5.0
* New: Add changelog for v3.9 to CHANGELOG.md

## 3.9 _(2019-03-08)_
* Fix: Default 'this_post' shortcode attribute to 1 instead of 0, since unlike widgets, shortcodes generally appear within the context of a post
* Fix: Call `wpdb::prepare()` with the proper number of arguments depending on context
* Change: Update shortcode builder widget to 005:
    * Don't show shortcode builder metabox within context of block editor
    * Add `show_metabox()`
* Change: Update widget to 012:

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Get_Custom_Field_Values
 */

ini_set( 'display_errors', 'on' );
error_reporting(E_ALL);

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/get-custom-field-values.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Get_Custom_Field_Values
 */

define( 'GET_CUSTOM_FIELD_VALUES_PLUGIN_DIR',  dirname( __FILE__, 3 ) );
define( 'GET_CUSTOM_FIELD_VALUES_PLUGIN_FILE', GET_CUSTOM_FIELD_VALUES_PLUGIN_DIR . '/get-custom-field-values.php' );

$polyfill_path = GET_CUSTOM_FIELD_VALUES_PLUGIN_DIR . '/vendor/yoast/phpunit-polyfills/phpunitpolyfills-autoload.php';
if ( file_exists( $polyfill_path ) ) {
	require $polyfill_path;
} else {
	echo "Error: PHPUnit Polyfills dependency not found.\n";
	echo "Run: composer require --dev yoast/phpunit-polyfills:\"^2.0\"\n";
	exit;
}

! defined( 'WP_RUN_CORE_TESTS' ) && define( 'WP_RUN_CORE_TESTS', false );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require GET_CUSTOM_FIELD_VALUES_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/tests/widget.php
<?php

defined( 'ABSPATH' ) or die();

class Get_Custom_Field_Values_Widget_Test extends WP_UnitTestCase {

	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function create_post_with_meta( $metas = array(), $post_data = array(), $make_global = false ) {
		$post_id = $this->factory->post->create( $post_data );

		if ( ! $metas ) {
			$metas = $this->get_sample_metas();
		}

		foreach ( $metas as $key => $val ) {
			if ( is_array( $val ) ) {
				foreach ( $val as $v ) {
					add_post_meta( $post_id, $key, $v );
				}
			} else {
				add_post_meta( $post_id, $key, $val );
			}
		}

		if ( $make_global ) {
			global $post;
			$post = get_post( $post_id );
		}

		return $post_id;
	}

	private function get_sample_metas() {
		return array(
			'mood'  => 'happy',
			'child' => array( 'adam', 'bob', 'cerise', 'diane' ),
			'color' => array( 'blue', 'white' ),
			'tshirt size' => 'M',
			'location' => 'Denver, CO',
			'Website' => 'example.com',
		);
	}

	private function widget_init( $config = array() ) {
		$post_id = $this->create_post_with_meta();

		c2c_GetCustomWidget::register_widget();
		$widget = new c2c_GetCustomWidget( 'abc_abc', '', array() );

		$default_config = array();
		foreach ( $widget->get_config() as $key => $val ) {
			$default_config[ $key ] = $val['default'];
		}
		$config = array_merge( $default_config, $config );

		if ( true === $config['post_id'] ) {
			$config['post_id'] = $post_id;
		}

		$settings = array( 'before_title' => '', 'after_title' => '', );

		return array( $post_id, $widget, $config, $settings );
	}


	//
	//
	// TESTS
	//
	//


	public function test_widget_class_exists() {
		$this->assertTrue( class_exists( 'c2c_GetCustomWidget' ) );
	}

	public function test_widget_version() {
		$this->assertEquals( '012', c2c_GetCustomWidget::version() );
	}

	public function test_widget_base_class_name() {
		$this->assertTrue( class_exists( 'c2c_Widget_013' ) );
	}

	public function test_widget_framework_version() {
		$this->assertEquals( '013', c2c_Widget_013::version() );
	}

	public function test_widget_hooks_widgets_init() {
		$this->assertEquals( 10, has_filter( 'widgets_init', array( 'c2c_GetCustomWidget', 'register_widget' ) ) );
	}

	public function test_widget_made_available() {
		$this->assertArrayHasKey( 'c2c_GetCustomWidget', $GLOBALS['wp_widget_factory']->widgets );
	}

	public function test_widget_body_with_class() {
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'class' => 'abcd', 'post_id' => true )  );

		$this->assertEquals( '<span class="abcd">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_id() {
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'id' => 'myid', 'post_id' => true ) );

		$this->assertEquals( '<span id="myid">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_class_and_id() {
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'class' => 'abcd', 'id' => 'myid', 'post_id' => true ) );

		$this->assertEquals( '<span id="myid" class="abcd">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_class_and_id_but_no_meta_value() {
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'nonexistent', 'class' => 'abcd', 'id' => 'myid', 'post_id' => true ) );

		$this->assertEmpty( $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_post_id_of_current_in_invalid_situation() {
		$p_id = $this->create_post_with_meta( array( 'mood' => 'confused' ) );

		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'post_id' => 'current' ) );

		$this->assertEmpty( $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_post_id_of_current() {
		$p_id = $this->create_post_with_meta( array( 'mood' => 'perplexed' ) );

		// Simulate conditions when it is valid to run.
		query_posts( array( 'p' => $p_id ) );

		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'post_id' => 'current' ) );

		$this->assertEquals( 'perplexed', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_explicit_post_id() {
		$p_id = $this->create_post_with_meta( array( 'mood' => 'joyous' ) );

		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'post_id' => $p_id ) );

		query_posts( array( 'p' => $post_id ) );

		$this->assertEquals( 'joyous', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_no_post_id() {
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood' ) );

		query_posts( array( 'p' => $post_id ) );

		$this->assertEquals( 'happy', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_no_post_id_and_no_current_post() {
		$post_id = $this->create_post_with_meta( array( 'mood' => 'befuddled' ) );
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'between' => ' and ' ) );

		$this->assertEquals( 'befuddled and happy', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_html_in_meta() {
		$post_id = $this->create_post_with_meta( array( 'concert' => '<sCript>alert(document.domain)</sCript>' ) );
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'concert', 'before' => '<strong>', 'after' => '</strong>', 'none' => 'None' ) );

		$this->assertEquals( "<strong>alert(document.domain)</strong>", $widget->widget_body( $widget->validate( $config ), '', $settings ) );
	}

	public function test_widget_with_html_in_fields() {
		$post_id = $this->create_post_with_meta( array( 'moodx' => '', [], true ) );
		$script = 'hello<sCript>alert(document.domain)</sCript>';
		$safe_script = '<strong>None found.</strong>';
		list( $post_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'moodx', 'before' => $script, 'after' => $script, 'none' => $safe_script ) );

		$this->assertEquals( "helloalert(document.domain)${safe_script}helloalert(document.domain)", $widget->widget_body( $widget->validate( $config ), '', $settings ) );
	}

}


get-custom.shortcode.php
<?php
/**
 * Get Custom Field Values plugin shortcode code.
 *
 * Copyright (c) 2004-2023 by Scott Reilly (aka coffee2code)
 *
 * @package c2c_GetCustomFieldValuesShortcode
 * @author  Scott Reilly
 * @version 008
 */

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_GetCustomFieldValuesShortcode' ) && class_exists( 'c2c_GetCustomWidget' ) ) :

class c2c_GetCustomFieldValuesShortcode {
	/**
	 * The instance of this object.
	 *
	 * @access public
	 * @since 007
	 * @var c2c_GetCustomFieldValuesShortcode
	 */
	public static $instance;

	public $name           = 'shortcode_get_custom_field_values';
	public $shortcode      = 'custom_field';
	public $title          = '';
	public $widget_handler = '';
	public $widget_base    = '';

	/**
	 * Returns version of the class.
	 *
	 * @since 003
	 *
	 * @return string
	 */
	public static function version() {
		return '008';
	}

	/**
	 * Registers the shortcode.
	 *
	 * @since 004
	 */
	public static function register() {
		self::$instance = new self( $GLOBALS['wp_widget_factory']->widgets['c2c_GetCustomWidget'] );
	}

	/**
	 * Constructor.
	 */
	public function __construct( $widget_handler ) {
		$this->title          = __( 'Get Custom Field Values - Shortcode', 'get-custom-field-values' );
		$this->widget_handler = $widget_handler;
		$this->widget_base    = 'widget-' . $this->widget_handler->id_base;
		$this->shortcode      = apply_filters( 'c2c_get_custom_field_values_shortcode', $this->shortcode );

		add_shortcode( $this->shortcode,         array( $this, 'shortcode' ) );
		add_action( 'load-post.php',             array( $this, 'register_post_page_hooks' ) );
		add_action( 'load-post-new.php',         array( $this, 'register_post_page_hooks' ) );
		add_filter( 'default_hidden_meta_boxes', array( $this, 'default_hidden_meta_boxes' ), 10, 2 );
	}

	/**
	 * Filters/actions to hook on the admin post.php page.
	 *
	 * @since 3.4
	 */
	public function register_post_page_hooks() {
		add_action( 'admin_print_footer_scripts', array( $this, 'admin_js' ) );
		add_action( 'do_meta_boxes',              array( $this, 'do_meta_box' ), 10, 3 );
	}

	/**
	 * Determines if post author can use shortcodes.
	 *
	 * @since 4.0
	 *
	 * @param null|int|WP_User $user The user ID, object, or null to refer to
	 *                               current post's author. Default null.
	 * @param null|int|WP_Post $post The post ID, object, or null to refer to
	 *                               current post. Default null.
	 * @return bool True if post author can use shortcode, else false.
	 */
	public function can_author_use_shortcodes( $user = null, $post = null ) {
		// Default to allowing shortcode usage as shortcode may appear in non-post contexts.
		// It is assumed those other contexts are trusted.
		$can = true;

		// If post not provided, use global post.
		$post = get_post( $post );

		// If user provided, get its object. Else, assume current post's author.
		if ( $user ) {
			if ( is_int( $user ) ) {
				$user = get_userdata( (int) $user );
			}
		} else {
			if ( $post ) {
				$user = get_userdata( $post->post_author );
			}
		}

		// User must have 'unfiltered_html' capability, which requires a user to
		// be known.
		if ( $user instanceof WP_User ) {
			$can = user_can( $user->ID, 'unfiltered_html' );
		}

		/**
		 * Filters if a given author (if applicable given the shortcode context),
		 * can use the 'custom_field' shortcode, or to have their use of it
		 * evaluated.
		 *
		 * Note: In contexts where no author is relevant (such as a shortcode
		 * appearing in a sidebar widget), then use of the shortcode is
		 * assumed to be allowed and trusted.
		 *
		 * @since 4.0
		 *
		 * @param bool          $can  Is shortcode use permitted?
		 * @param WP_User|false $user User object, or false if no user in context.
		 * @param WP_Post|false $post Post object, or false if no post in context.
		 */
		return (bool) apply_filters( 'get_custom_field_values/can_author_use_shortcodes', $can, $user, $post );
	}

	/**
	 * Determines if the metabox should be shown.
	 *
	 * The metabox should only be shown in the classic editor.
	 *
	 * @since 005
	 * @since 4.0 Prevent users who cannot publish posts from seeing the metabox.
	 *
	 * @return bool True if the metabox can be shown, false otherwise.
	 */
	public function show_metabox() {
		$current_screen = get_current_screen();
		$show = true;

		// Ensure current screen is for a post.
		if ( $show ) {
			$show = 'post' === $current_screen->base;
		}

		// Ensure block editor is not enabled (metabox is not compatible in that context).
		if ( $show ) {
			$show = ! method_exists( $current_screen, 'is_block_editor' ) || ! $current_screen->is_block_editor();
		}

		// Ensure post author can use shortcodes.
		if ( $show ) {
			/**
			 * Filters if the shortcode builder metabox is shown.
			 *
			 * @ssince 4.0
			 *
			 * @param bool $show Show the metabox? Default is whether the post author can publish unfiltered HTML.
			 */
			$show = (bool) apply_filters( 'get_custom_field_values/show_metabox', $this->can_author_use_shortcodes() );
		}

		return $show;
	}

	/**
	 * Register meta box
	 *
	 * By default, the shortcode builder is present for all post types. Filter
	 * 'c2c_get_custom_field_values_post_types' to limit its use.
	 *
	 * @since 3.4
	 *
	 * @param string  $post_type The post type
	 * @param string  $type      The mode for the meta box (normal, advanced, or side)
	 * @param WP_Post $post      The post
	 */
	public function do_meta_box( $post_type, $type, $post ) {
		if ( ! $this->show_metabox() ) {
			return;
		}

		$post_types = (array) apply_filters( 'c2c_get_custom_field_values_post_types', get_post_types() );
		if ( ! in_array( $post_type, $post_types ) ) {
			return;
		}

		add_meta_box( $this->name, $this->title, array( $this, 'form' ), $post_type, 'side', 'core' );
	}

	/**
	 * Hides the shortcode wizard by default.
	 *
	 * @since 3.5
	 *
	 * @param array     $hidden List of all hidden metaboxes
	 * @param WP_Screen $screen Screen object.
	 */
	public function default_hidden_meta_boxes( $hidden, $screen ) {
		if ( ! in_array( $this->name, $hidden ) ) {
			$hidden[] = $this->name;
		}

		return $hidden;
	}

	/**
	 * Shortcode handler.
	 *
	 * @since 4.0 Prevent evaluation of shortcodes in posts author by users who cannot publish posts.
	 *
	 * @param array  $atts    The shortcode attributes
	 * @param string $content The text wrapped by the shortcode's opening and closing tags.
	 * @return string
	 */
	public function shortcode( $atts, $content = null ) {
		$defaults = array();

		if ( ! $this->can_author_use_shortcodes() ) {
			return '';
		}

		foreach ( $this->widget_handler->get_config() as $opt => $values ) {
			// Unlike widgets, shortcodes generally exist in the context of a post, so
			// override widget default.
			if ( 'this_post' === $opt ) {
				$values['default'] = true;
			}

			$defaults[ $opt ] = isset( $values['default'] ) ? $values['default'] : '';
		}

		$atts2 = shortcode_atts( $defaults, $atts );

		foreach ( array_keys( $this->widget_handler->get_config() ) as $key ) {
			$$key = $atts2[ $key ];
		}

		$ret = '';

		if ( $post_id ) {
			if ( 'current' == $post_id ) {
				$ret = c2c_get_current_custom( $field, $before, $after, $none, $between, $before_last );
			} elseif ( $random ) {
				$ret = c2c_get_random_post_custom( $post_id, $field, $limit, $before, $after, $none, $between, $before_last );
			} else {
				$ret = c2c_get_post_custom( $post_id, $field, $before, $after, $none, $between, $before_last );
			}
		} else {
			if ( $this_post ) {
				$ret = c2c_get_custom( $field, $before, $after, $none, $between, $before_last );
			} elseif ( $random ) {
				$ret = c2c_get_random_custom( $field, $before, $after, $none );
			} else {
				$ret = c2c_get_recent_custom( $field, $before, $after, $none, $between, $before_last, $limit );
			}
		}

		// If either 'id' or 'class' attribute was defined, then wrap output in span
		if ( $ret && ( $id || $class ) ) {
			$tag = '<span';
			if ( $id ) {
				$tag .= ' id="' . esc_attr( $id ) . '"';
			}
			if ( $class ) {
				$tag .= ' class="' . esc_attr( $class ) . '"';
			}
			$tag .= ">$ret</span>";
			$ret = $tag;
		}

		return $ret;
	}

	/**
	 * Outputs the JS for the shortcode wizard.
	 */
	public function admin_js() {
		if ( ! $this->show_metabox() ) {
			return;
		}

		echo <<<JS
		<script type="text/javascript">
			jQuery.noConflict();
			var {$this->name} = function () {}

			{$this->name}.prototype = {
				options           : {},
				generateShortCode : function() {
					var content = this['options']['content'];
					delete this['options']['content'];

					var attrs = '';
					jQuery.each(this['options'], function(name, value){
						if (value != '') {
							attrs += ' ' + name + '="' + value + '"';
						}
					});
					return '[{$this->shortcode}' + attrs + ' /]';
				},
				sendToEditor      : function(f) {
					var collection = jQuery(f).find("input[id^={$this->widget_base}]:not(input:checkbox), \
										input[id^={$this->widget_base}]:checkbox:checked, \
										textarea[id^={$this->widget_base}]");
					var \$this = this;
					collection.each(function () {
						var name = this.name.substring(this.name.lastIndexOf('[')+1, this.name.length-1);
						if (\$this['options'][name] == undefined)
							\$this['options'][name] = this.value;
						else
							\$this['options'][name] += ', '+this.value;
					});

					// Delete between and before_last values if a limit of 1 was specified (since there will be at most 1)
					if ( \$this['options']['limit'] == '1' ) {
						delete \$this['options']['between'];
						delete \$this['options']['before_last'];
					}

					send_to_editor(this.generateShortCode());
					/* Delete data after generating shortcode so that the form can be used to generate another shortcode */
					collection.each(function () {
						var name = this.name.substring(this.name.lastIndexOf('[')+1, this.name.length-1);
						delete \$this['options'][name];
					});
					return false;
				}
			}

			var admin_{$this->name} = new {$this->name}();
		</script>

JS;
	}

	/**
	 * Outputs shortcode wizard form.
	 */
	public function form() {
		$this->widget_handler->form( array(), array( 'title' ) );
		printf(
			'<p class="submit"><input type="button" class="button-primary" onclick="return admin_%s.sendToEditor(this.form);" value="%s" /></p>' . "\n",
			esc_attr( $this->name ),
			esc_attr__( 'Send shortcode to editor', 'get-custom-field-values' )
		);
	}

} // end class c2c_GetCustomFieldValuesShortcode

add_action( 'init', array( 'c2c_GetCustomFieldValuesShortcode', 'register' ), 11 );

endif; // end if !class_exists()
